### ISSUE TITLE
`sudo service cassandra stop` doesn't always

### ISSUE THREAD

**Issue by mirw** [Metaswitch/clearwater-cassandra#27]
We often tell users to issue `sudo service cassandra stop` to stop and restart Cassandra.  Unfortunately, `sudo service cassandra stop` always does **just** a clean termination - it doesn't force it if the process doesn't die.  This could be sensible (we don't want to lose data), but we need to consider what happens when this fails.

The specific instance I hit was that we asked Cassandra to shut down and it cycled as follows:

```
 INFO 14:54:36,968 Listening for thrift clients...
 INFO 14:54:36,977 Stop listening to thrift clients
 WARN 14:54:36,981 Transport error occurred during acceptance of message.
org.apache.thrift.transport.TTransportException: java.net.SocketException: Socket closed
        at org.apache.cassandra.thrift.TCustomServerSocket.acceptImpl(TCustomServerSocket.java:109)
        at org.apache.cassandra.thrift.TCustomServerSocket.acceptImpl(TCustomServerSocket.java:36)
        at org.apache.thrift.transport.TServerTransport.accept(TServerTransport.java:31)
        at org.apache.cassandra.thrift.CustomTThreadPoolServer.serve(CustomTThreadPoolServer.java:110)
        at org.apache.cassandra.thrift.ThriftServer$ThriftServerThread.run(ThriftServer.java:111)
Caused by: java.net.SocketException: Socket closed
        at java.net.PlainSocketImpl.socketAccept(Native Method)
        at java.net.AbstractPlainSocketImpl.accept(AbstractPlainSocketImpl.java:375)
        at java.net.ServerSocket.implAccept(ServerSocket.java:478)
        at java.net.ServerSocket.accept(ServerSocket.java:446)
        at org.apache.cassandra.thrift.TCustomServerSocket.acceptImpl(TCustomServerSocket.java:103)
        ... 4 more
 WARN 14:54:36,982 Transport error occurred during acceptance of message.
org.apache.thrift.transport.TTransportException: java.net.SocketException: Socket is closed
        at org.apache.cassandra.thrift.TCustomServerSocket.acceptImpl(TCustomServerSocket.java:109)
        at org.apache.cassandra.thrift.TCustomServerSocket.acceptImpl(TCustomServerSocket.java:36)
        at org.apache.thrift.transport.TServerTransport.accept(TServerTransport.java:31)
        at org.apache.cassandra.thrift.CustomTThreadPoolServer.serve(CustomTThreadPoolServer.java:110)
        at org.apache.cassandra.thrift.ThriftServer$ThriftServerThread.run(ThriftServer.java:111)
Caused by: java.net.SocketException: Socket is closed
        at java.net.ServerSocket.accept(ServerSocket.java:442)
        at org.apache.cassandra.thrift.TCustomServerSocket.acceptImpl(TCustomServerSocket.java:103)
        ... 4 more
```

After about 30 minutes, I spotted that it wasn't coming up, logged in and "vi"-ed the log file.  This was so big that the OOM killer killed Cassandra, and everything sprung into life.

One solution here might be to fix https://github.com/Metaswitch/clearwater-monit/issues/15, but even this relies on Cassandra getting far enough through termination processing to close port 9160.

**Comment by mirw** [comment]
I think I've seen this causing clearwater-etcd-based clusters not to form - raising to high priority.

**Comment by bossmc** [comment]
https://github.com/Metaswitch/clearwater-etcd/issues/80 bypasses this issue by killing Cassandra explicitly after stopping the service.  Lowering the propriety now that this doesn't block initial spin up of a deployment.

**Comment by rkd-msw** [comment]
It looks like "service cassandra stop" does kill (SIGKILL) Cassandra if it hasn't terminated (SIGTERM) after 30 seconds - see https://github.com/Metaswitch/clearwater-cassandra/blob/master/clearwater-cassandra/etc/init.d/cassandra.clearwater#L113.

**Comment by rkd-msw** [comment]
Digging into this a bit more, we moved from Cassandra 1.2.12 to 2.0.14 a few weeks after this issue was raised (https://github.com/Metaswitch/clearwater-cassandra/commit/0d63f7a7b47383af75d0eb946e5958506f683122) and 1.2.12 had different behaviour without the SIGKILL (https://github.com/apache/cassandra/blob/cassandra-1.2.12/debian/init#L169). That means I understand why we originally hit this iussue, and I'm confident that this is fixed (i.e. it's not the case that we hit this issue even with the SIGKILL behaviour I linked to above).
